OpenRA.Mods.Common/Activities/Attack.cs

```diff
@@ -72,27 +72,63 @@ public override Activity Tick(Actor self)
 			return NextActivity;
 		}
 
+		protected bool ValidateFrozenActorTarget(Actor self, AttackBase attack)
+		{
+			// If target was a frozen actor got replaced with an actor, try switching
+			if (!target.IsValidFor(self))
+			{
+				if (target.FrozenActor != null && target.FrozenActor.Actor != null)
+				{
+					var loc = self.World.Map.CellContaining(target.FrozenActor.CenterPosition);
+					foreach (var actor in self.World.ActorMap.GetActorsAt(loc))
+					{
+						if (actor.ActorID == target.FrozenActor.ID && self.Owner.CanTargetActor(actor))
+						{
+							target = Target.FromActor(actor);
+							return true;
+						}
+					}
+				}
+
+				return false;
+			}
+
+			var positionableLoc = self.World.Map.CellContaining(target.CenterPosition);
+			if (attack.Info.AttackRequiresEnteringCell && !positionable.CanEnterCell(positionableLoc, null, false))
```

Comment:
This could then be pulled out into `InnerTick`, after the (frozen-)actor target validation.